<a href="https://colab.research.google.com/github/dSudario9/Modelos-Supervisionados-Diversos/blob/main/Tunning_de_Hiperparametros_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 **Classificação de saúde fetal.
 Nessa analise  vamos procurar o melhor classificador, aquel que tiver o melhor rsultado será o nosso modelo predefinido. 

In [ ]:
# Import das Bibliotecas 
import pandas as pd

EDA

In [ ]:
# import base de dados 
df = pd.read_csv("/content/fetal_health.csv")

In [ ]:
# Visualizar base de dados 
df.head()

baseline value  accelerations  ...  histogram_tendency  fetal_health
0           120.0          0.000  ...                 1.0           2.0
1           132.0          0.006  ...                 0.0           1.0
2           133.0          0.003  ...                 0.0           1.0
3           134.0          0.003  ...                 1.0           1.0
4           132.0          0.007  ...                 1.0           1.0

[5 rows x 22 columns]

In [ ]:
# Linha e Colunas 
df.shape

(2126, 22)

In [ ]:
# Definindo Variavel Target  alvo 
X = df.drop(columns=["fetal_health"])
Y = df["fetal_health"]

**Separando dados em Train e Test **

In [ ]:
# Import de Biblioteca do Modelo e função Train_test_split.

from sklearn.model_selection import train_test_split

In [ ]:
# Separando 205 do dataset para test 
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.2)

In [ ]:
print(f'X_train{X_train.shape}')
print(f'X_test{X_test.shape}')
print(f'y_train{y_train.shape}')
print(f'y_test{y_test.shape}')




X_train(1700, 21)
X_test(426, 21)
y_train(1700,)
y_test(426,)


*** Criando Classificador RandomForestClassifier***

In [ ]:
# import 
from sklearn.ensemble import RandomForestClassifier

In [ ]:
# Classificador 
clf = RandomForestClassifier()


***Treinando Modelo ***

In [ ]:
clf.fit(X_train, y_train)

RandomForestClassifier()

In [ ]:
# Predição 
y_pred = clf.predict(X_test)

Metrica F1 SCORE 

In [ ]:
# Import metrica 
from sklearn.metrics import f1_score

In [ ]:
f1_score(y_test,y_pred, average='weighted')*100

93.48180035497539

In [ ]:
# Ultilizando Metodo GRID SEARCH  para buscar um melhor parametro


In [ ]:
# Import 
from sklearn.model_selection import GridSearchCV

In [ ]:
# Força nossos parametros a buscar as  melhores metricas criando um dicionario com os parametros da busca. 
params_grid = {
    "n_estimators":[100, 150, 200],
    "criterion":["gini", "entropy"],
    "max_depth":[None, 4, 5],
    "min_samples_split":[2, 3, 4],
    "max_features":['auto', 5, 8],
    "bootstrap":[False, True]
    
}

In [ ]:
# Selecionando Classificador 

clf = RandomForestClassifier()

# Fazendo Busca de Imperparametros 

grid_search = GridSearchCV(clf, param_grid=params_grid, cv=3, n_jobs=-1, verbose=2)

# Selecionando os melhores parametros

grid_search.fit(X_train, y_train)

Fitting 3 folds for each of 324 candidates, totalling 972 fits


GridSearchCV(cv=3, estimator=RandomForestClassifier(), n_jobs=-1,
             param_grid={'bootstrap': [False, True],
                         'criterion': ['gini', 'entropy'],
                         'max_depth': [None, 4, 5],
                         'max_features': ['auto', 5, 8],
                         'min_samples_split': [2, 3, 4],
                         'n_estimators': [100, 150, 200]},
             verbose=2)

In [ ]:
# Visualizando os melhores parametros encontrados. 
grid_search.best_params_

{'bootstrap': True,
 'criterion': 'entropy',
 'max_depth': None,
 'max_features': 8,
 'min_samples_split': 3,
 'n_estimators': 200}

In [ ]:
# # Visualizar o melhor Classificador estimador 

clf_best = grid_search.best_estimator_

In [ ]:
# Predição 
y_pred = clf_best.predict(X_test)

In [ ]:
# Visualizando Resultado Metrica 
f1_score(y_test, y_pred, average="weighted")*100

94.48735295481333

**Observando o resultado da metrica com o  Imperparametro  conseguimos melhorar a performace  de 93.48180035497539%  para 94.48735295481333% .Podemos  realizar um Ajuste fino nos parametros acima no grid para melhorar nossa performace e encontrar um melhor resultado .**

Vamos testar uma outra Tecnica conhecida como BayesSearchCV para tentar customizar o modelo de uma melhor forma para simplificar nosso algoritio
e afunilzar as buscar de imperparametros para nao precisar passar por todas as combinações 

In [ ]:
# install da Biblioteca  scikit-optimize

!pip install scikit-optimize

     |████████████████████████████████| 100 kB 5.9 MB/s 


In [ ]:
# import da Biblioteca BayesSearchCV 

from skopt import BayesSearchCV

In [ ]:
# Criando Campo de busca com parametros . 
params_grid = {
    "n_estimators":[100, 150, 200],
    "criterion":["gini", "entropy"],
    "max_depth":[None, 4, 5],
    "min_samples_split":[2, 3, 4],
    "max_features":[3, 5, 8],
    "bootstrap":[False, True]
    
}

In [ ]:
# Construindo  Classificador 
clf = RandomForestClassifier()

## Fazendo Busca de Imperparametros 

bayes_search = BayesSearchCV(clf, search_spaces=params_grid, cv=3, n_jobs=1,verbose=2)

# Selecionando os melhores parametros. 
bayes_search.fit(X_train, y_train)

Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=True, criterion=entropy, max_depth=5, max_features=3, min_samples_split=4, n_estimators=200; total time=   0.4s
[CV] END bootstrap=True, criterion=entropy, max_depth=5, max_features=3, min_samples_split=4, n_estimators=200; total time=   0.4s
[CV] END bootstrap=True, criterion=entropy, max_depth=5, max_features=3, min_samples_split=4, n_estimators=200; total time=   0.4s
Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=True, criterion=gini, max_depth=5, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.4s
[CV] END bootstrap=True, criterion=gini, max_depth=5, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.4s
[CV] END bootstrap=True, criterion=gini, max_depth=5, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.4s
Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=True, criterion=entropy, ma

/usr/local/lib/python3.7/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "


Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=100; total time=   0.4s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=100; total time=   0.4s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=100; total time=   0.4s
Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END boot

/usr/local/lib/python3.7/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "


Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=False, criterion=entropy, max_depth=4, max_features=3, min_samples_split=3, n_estimators=150; total time=   0.3s
[CV] END bootstrap=False, criterion=entropy, max_depth=4, max_features=3, min_samples_split=3, n_estimators=150; total time=   0.3s
[CV] END bootstrap=False, criterion=entropy, max_depth=4, max_features=3, min_samples_split=3, n_estimators=150; total time=   0.3s
Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=Fal

/usr/local/lib/python3.7/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "


Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s


/usr/local/lib/python3.7/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "


Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=3, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=3, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=3, n_estimators=200; total time=   0.7s


/usr/local/lib/python3.7/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "


Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=True, criterion=entropy, max_depth=5, max_features=8, min_samples_split=2, n_estimators=200; total time=   0.6s
[CV] END bootstrap=True, criterion=entropy, max_depth=5, max_features=8, min_samples_split=2, n_estimators=200; total time=   0.6s
[CV] END bootstrap=True, criterion=entropy, max_depth=5, max_features=8, min_samples_split=2, n_estimators=200; total time=   0.6s


/usr/local/lib/python3.7/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "


Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=2, n_estimators=200; total time=   0.7s


/usr/local/lib/python3.7/dist-packages/skopt/optimizer/optimizer.py:449: UserWarning: The objective has been evaluated at this point before.
  warnings.warn("The objective has been evaluated "


Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=4, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=4, n_estimators=200; total time=   0.7s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=4, n_estimators=200; total time=   0.7s
Fitting 3 folds for each of 1 candidates, totalling 3 fits
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=3, n_estimators=150; total time=   0.6s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=3, n_estimators=150; total time=   0.5s
[CV] END bootstrap=False, criterion=entropy, max_depth=None, max_features=5, min_samples_split=3, n_estimators=150; total time=   0.5s


BayesSearchCV(cv=3, estimator=RandomForestClassifier(),
              search_spaces={'bootstrap': [False, True],
                             'criterion': ['gini', 'entropy'],
                             'max_depth': [None, 4, 5],
                             'max_features': [3, 5, 8],
                             'min_samples_split': [2, 3, 4],
                             'n_estimators': [100, 150, 200]},
              verbose=2)

In [ ]:
# Vizualizando os melhores parametros encontrados. 
bayes_search.best_params_

OrderedDict([('bootstrap', False),
             ('criterion', 'entropy'),
             ('max_depth', None),
             ('max_features', 5),
             ('min_samples_split', 2),
             ('n_estimators', 200)])

In [ ]:
# Aqui atribuimos  os melhoes estimadores  com o parametro
bayes_best = bayes_search.best_estimator_

In [ ]:
y_pred = bayes_best.predict(X_test)

In [ ]:
f1_score(y_test, y_pred, average="weighted")*100

94.24504115646648

 **Nesse  imperparametro tivemos uma metrica de 94.24504115646648% podendo assim mudar os parametros afunilando o algoritimo para que possamos chegar a um melhor resultado. 

Esse Foi o nosso tunning de Imperparametros**